From 4cd8fa85497db24dcf59d3148748575aa8e3c7a2 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Mon, 3 Dec 2012 16:13:49 +0900
Subject: [PATCH 82/97] ext/dfbvideosink: calculate Y plane stride when the
 color format is planar or semi planar

src_pitch is determined by dividing the buffer size by the image height, which
is correct when a packed format is specified. However, When the planar or
the semi planar fomats are specified, src_picth (for the Y plane)
must be adjusted to account for the CbCr plane that is present in the same
buffer.
---
 ext/directfb/dfbvideosink.c |   19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git a/ext/directfb/dfbvideosink.c b/ext/directfb/dfbvideosink.c
index 3f5db8d..547a13f 100644
--- a/ext/directfb/dfbvideosink.c
+++ b/ext/directfb/dfbvideosink.c
@@ -2257,8 +2257,9 @@ gst_dfbvideosink_show_frame (GstBaseSink * bsink, GstBuffer * buf)
     gint dest_pitch, src_pitch;
     GstStructure *structure;
     gint line;
+    DFBSurfacePixelFormat src_format;
 #if defined(HAVE_SHVIO)
-    DFBSurfacePixelFormat src_format, dst_format;
+    DFBSurfacePixelFormat dst_format;
     guint8 *src_datay, *src_datac;
 #endif
 
@@ -2348,7 +2349,6 @@ gst_dfbvideosink_show_frame (GstBaseSink * bsink, GstBuffer * buf)
     /* Source video rowbytes */
     src_pitch = GST_BUFFER_SIZE (buf) / src.h;
 
-#if defined(HAVE_SHVIO)
     src_format = gst_dfbvideosink_get_format_from_caps (GST_BUFFER_CAPS (buf));
     if (src_format == DSPF_UNKNOWN) {
       GST_WARNING_OBJECT (dfbvideosink,
@@ -2356,6 +2356,21 @@ gst_dfbvideosink_show_frame (GstBaseSink * bsink, GstBuffer * buf)
       return GST_FLOW_UNEXPECTED;
     }
 
+    /* calculate a Y plane stride when the color format is planar or
+       semi planar */
+    switch (src_format) {
+      case DSPF_NV12:
+      case DSPF_YV12:
+        src_pitch = src_pitch * 2 / 3;
+        break;
+      case DSPF_NV16:
+        src_pitch /= 2;
+        break;
+      default:
+        break;
+    }
+
+#if defined(HAVE_SHVIO)
     if (dfbvideosink->interlaced) {
       gint is_src_odd, is_dst_odd;
 
-- 
1.7.10.4

